package/MDAnalysis/analysis/dihedrals.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import warnings
+
+from MDAnalysis.analysis.base import AnalysisBase
+
+def dihedral_calc(atomgroups):
+    """Calculates phi and psi angles for a list of AtomGroups over trajectory.
+
+    Parameters
+    ----------
+    atomgroups : list of AtomGroups
+        must be a list of one or more AtomGroups containing 5 atoms in the
+        correct order (i.e. C-N-CA-C-N)
+
+    Returns
+    -------
+    angles : numpy.ndarray
+        An array of time steps which contain (phi,psi) for all AtomGroups.
+    """
+
+    dihedrals = [atomgroup.dihedral for atomgroup in atomgroups]
+    angles = [dih.value() for dih in dihedrals]
+
+    return angles
+
+
+class Ramachandran(AnalysisBase):
+    """Calculate phi and psi dihedral angles of specified residues.
+
+    Note
+    ----
+    Run the analysis with :meth:`Ramachandran.run()`, which stores the results
+    in the array :attr:`Ramachandran.angles`. A axes object can be obtained
+    with :meth: `Ramachandran.run().plot()`.
+
+    If the residue selection is beyond the scope of the protein, then an error
+    will be raised. If the residue selection includes the first or last residue
+    then a warning will be raised, and the final array of angles will not
+    include those residues.
+
+    """
+    def __init__(self, atomgroup, **kwargs):
+        r"""Parameters
+        ----------
+        atomgroup : Atomgroup
+            atoms for residues for which phi and psi are calculated
+        start : int, optional
+            starting frame, default None becomes 0.
+        stop : int, optional
+            Frame index to stop analysis. Default: None becomes
+            n_frames. Iteration stops *before* this frame number,
+            which means that the trajectory would be read until the end.
+        step : int, optional
+            step between frames, default None becomes 1.
+
+        """
+        super(Ramachandran, self).__init__(atomgroup.universe.trajectory, **kwargs)
+        self.atomgroup = atomgroup
+
+    def _prepare(self):
+        self.residues = self.atomgroup.residues
+        res_min = np.min(self.atomgroup.universe.select_atoms("protein").residues)
+        res_max = np.max(self.atomgroup.universe.select_atoms("protein").residues)
+        if any([(residue < res_min or residue > res_max) for residue in self.residues]):
+            raise IndexError("Selection exceeds protein length")
```

Comment:
how is this possible? Literally how. I would need to get an atom that is not in the universe topology. 

I like to check things but I can't think of a case where this would ever be triggered.

Reply:
if `self.atomgroup.residues` contains residues which aren't in `protein` I think?

Reply:
Solvent molecules are listed as residues in the topology files (at least in the .gro files I've been using) so it possible to have a residue group that contains solvent molecules. The idea was to give more feedback on what the problem actually is rather than being confused when `phi_selection()` or `psi_selection()` returned None.

Reply:
But using `ag.issubset(protein)` is easier and more robust for the test anyway.

Reply:
Thanks for the comment. I agree with @orbeckst to use set-semantics because it's easier to understand and more robust. I would also update the warning message to `Found atoms outside of protein. Only atoms inside of a 'protein' selection can be used to calculate dihedrals`.